packages/cli/src/commands/hydrogen/upgrade.ts

```diff
@@ -0,0 +1,2931 @@
+  const currentRemix =
+    Object.entries(currentDependencies).find(isRemixDependency);
+  const selectedRemix = Object.entries(selectedRelease.dependencies).find(
+    isRemixDependency,
+  );
+
+  if (currentRemix && selectedRemix) {
+    const shouldUpgradeRemix = semver.lt(
+      getAbsoluteVersion(currentRemix[1]),
+      getAbsoluteVersion(selectedRemix[1]),
+    );
+
+    if (shouldUpgradeRemix) {
+      args.push(
+        ...appendRemixDependencies({currentDependencies, selectedRemix}),
+      );
+    }
+  }
+
+  return args;
+}
+
+/**
+ * Installs the new Hydrogen dependencies
+ */
+export async function upgradeNodeModules({
+  appPath,
+  selectedRelease,
+  currentDependencies,
+}: {
+  appPath: string;
+  selectedRelease: Release;
+  currentDependencies: Dependencies;
+}) {
+  await renderTasks(
+    [
+      {
+        title: `Upgrading dependencies`,
+        task: async () => {
+          await installNodeModules({
+            directory: appPath,
+            packageManager: await getPackageManager(appPath),
+            args: buildUpgradeCommandArgs({
+              selectedRelease,
+              currentDependencies,
+            }),
+          });
+        },
+      },
+    ],
+    {},
+  );
+}
+
+/**
+ * Appends the current @remix-run dependencies to the upgrade command
+ */
+function appendRemixDependencies({
+  currentDependencies,
+  selectedRemix,
+}: {
+  currentDependencies: Dependencies;
+  selectedRemix: [string, string];
+}) {
+  const command: string[] = [];
+  for (const [name, version] of Object.entries(currentDependencies)) {
+    const isRemixPackage = isRemixDependency([name, version]);
+    if (!isRemixPackage) {
+      continue;
+    }
+    command.push(`${name}@${getAbsoluteVersion(selectedRemix[1])}`);
+  }
+  return command;
+}
+
+/**
+ * Gets the absolute version from a pinned or unpinned version
+ */
+export function getAbsoluteVersion(version: string) {
+  const isPinned = /^[\d\.]+$/.test(version);
+  const currentPinnedVersion = isPinned ? version : version.slice(1);
+  return currentPinnedVersion;
+}
+
+/**
+ * Prompts the user to select a version from the list of available upgrades
+ */
+async function promptUpgradeOptions(
+  currentVersion: string,
+  upgradesAvailable: Array<Release>,
+) {
+  if (!upgradesAvailable?.length) {
+    return null;
+  }
+
+  // Build the list of upgrade options to display to the user
+  const choices = upgradesAvailable.map((release, index) => {
+    const {version, title} = release;
+    let isLatest = false;
+
+    if (index === 0) {
+      isLatest = true;
+    }
+
+    // TODO: add group sorting function to cli-kit select prompt
+    const majorVersion = `${semver.major(version)}.${semver.minor(version)}`;
+
+    const isFirstMajorVersion = semver.patch(version) === 0;
+
+    return {
+      // group: majorVersion,
+      label: `${version} ${
+        isLatest ? '(latest)' : isFirstMajorVersion ? '(major)' : ''
+      } - ${title}`,
+      value: version,
```

Comment:
Maybe this could be simplified too `..., value: release } as Choice<Release>` instead?